Add SIG Node spotlight interview (#498)
kubernetes/contributor-site: content/en/blog/2024/sig-node-spotlight/sig-node-spotlight.md

---
layout: blog
title: "Spotlight on SIG Node"
slug: sig-node-spotlight-2024

> date: 2024-05-09

Reviewer: Let's push this back. How about the 21st of May?

Author: Perfect, will update in the next push.

author: "Arpit Agrawal"
---

> In the world of container orchestration, [Kubernetes](https://kubernetes.io/docs/home/) reigns

Reviewer: ```suggestion
In the world of container orchestration, [Kubernetes](https://kubernetes.io/) reigns
```

supreme, powering some of the most complex and dynamic applications across the globe. Behind the
scenes, a network of Special Interest Groups (SIGs) drives Kubernetes' innovation and stability.

> Today, I have the privilege of speaking with [Matthias
> Bertschy](https://www.linkedin.com/in/matthias-bertschy-b427b815/), [Gunju
> Kim](https://www.linkedin.com/in/gunju-kim-916b33190/), and [Sergey
> Kanzhelev](https://www.linkedin.com/in/sergeykanzhelev/), members of [SIG
> Node](https://github.com/kubernetes/community/blob/master/sig-node/README.md), who will shed some
> light on their roles, challenges, and the exciting developments within SIG Node.

Reviewer: (nit)
```suggestion
Today, I have the privilege of speaking with
[Matthias Bertschy](https://www.linkedin.com/in/matthias-bertschy-b427b815/),
[Gunju Kim](https://www.linkedin.com/in/gunju-kim-916b33190/), and
[Sergey Kanzhelev](https://www.linkedin.com/in/sergeykanzhelev/), members of
[SIG Node](https://github.com/kubernetes/community/blob/master/sig-node/README.md),
who will shed some light on their roles, challenges, and the exciting developments within SIG Node.
```

We like to avoid line breaks within hyperlinks.

_Answers given collectively by all interviewees will be marked by their initials._

## Introductions

**Arpit:** Thank you for joining us today. Could you please introduce yourselves and provide a brief
overview of your roles within SIG Node?

**Matthias:** My name is Matthias Bertschy, I am French and live next to Lake Geneva, near the

> French Alps. I have been a Kubernetes contributor since 2017, a reviewer for SIG node and a

Reviewer 2: ```suggestion
French Alps. I have been a Kubernetes contributor since 2017, a reviewer for SIG Node and a
```

maintainer of [Prow](https://docs.prow.k8s.io/docs/overview/). I work as a Senior Kubernetes

> Developer for a security startup named [ARMO](https://www.armosec.io/) which donated

Reviewer 2: ```suggestion
Developer for a security startup named [ARMO](https://www.armosec.io/), which donated
```

[Kubescape](https://www.cncf.io/projects/kubescape/) to the CNCF.

![Lake Geneva and the Alps](/blog/2024/sig-node-spotlight/Lake_Geneva_and_the_Alps.jpg)

**Gunju:** My name is Gunju Kim. I am a software engineer at

> [NAVER](https://www.navercorp.com/naver/naverMain), where I focus on developing cloud cloud-serving

Reviewer 2: `cloud cloud-serving platform`? Can we clarify this, please?

platform for search services. I have been contributing to the Kubernetes project in my free time
since 2021.

**Sergey:** My name is Sergey Kanzhelev. I have worked on Kubernetes and [Google Kubernetes
Engine](https://cloud.google.com/kubernetes-engine) for 3 years and have worked on open-source
projects for many years now. I am a chair of SIG Node.

## Understanding SIG Node

**Arpit:** Thank you! Could you provide our readers with an overview of SIG Node's responsibilities
within the Kubernetes ecosystem?

> **M/G/S:** SIG Node is one of the first if not the very first SIG in Kubernetes. The SIG is

Reviewer 2: It was the first (or one of the first) SIG, or it is the first people interact with?

responsible for all iterations between Kubernetes and node resources, as well as node maintenance

> itself. This is quite a large scope, and SIG owns a large part of the Kubernetes codebase. Because

Reviewer 2: ```suggestion
itself. This is quite a large scope, and the SIG owns a large part of the Kubernetes codebase. Because
```

Storage, and SIG Security and almost any new features and developments in Kubernetes involves SIG
Node in some way.

**Arpit**: How does SIG Node contribute to Kubernetes' performance and stability?

**M/G/S:** Kubernetes works on nodes of many different sizes and shapes, from small physical VMs

> with cheap hardware to large AI/ML-optimized GPU-enabled nodes. Nodes may stay online for months or
> maybe be short-lived and be preempted at any moment as they are running on excess compute of a cloud
> provider.

Reviewer 3: ```suggestion
with cheap hardware to large AI/ML-optimized GPU-enabled nodes. Nodes may stay online for months or
be short-lived and preempted at any moment, as they are running on excess compute of a cloud
provider.
```

Reviewer: If we're quoting people, we usually keep their actual words.

> [Kubelet](https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet/) -- the
> Kubernetes agent on a node -- must work in all these environments reliably. As for the performance

Reviewer: ```suggestion
[`kubelet`](https://kubernetes.io/docs/concepts/overview/components/#kubelet) — the
Kubernetes agent on a node — must work in all these environments reliably. As for the performance
```

Kubernetes agent on a node -- must work in all these environments reliably. As for the performance

> of Kubelet operations, this is becoming increasingly important today. On one hand, as Kubernetes is

Reviewer 2: ```suggestion
of kubelet operations, this is becoming increasingly important today. On one hand, as Kubernetes is
```

> being used on extra small nodes more and more often in telecom and retail environments - it needs to

Reviewer 2: ```suggestion
being used on extra small nodes more and more often in telecom and retail environments, it needs to
```

> scale into the smallest footprint possible. On the other hand, with AI/ML workloads where every Node

Reviewer 2: ```suggestion
scale into the smallest footprint possible. On the other hand, with AI/ML workloads where every node
```

is extremely expensive, every second of delayed operations can visibly change the price of
computation.

> # Challenges and Opportunities

Reviewer 2: ```suggestion
## Challenges and Opportunities
```

**Arpit:** What upcoming challenges and opportunities is SIG Node keeping an eye on?

**M/G/S:** As Kubernetes enters the second decade of its life, we see a huge demand to support new

> workload types. And SIG Node will play a big role in this.The Sidecar KEP, which we will be talking

Reviewer 2: ```suggestion
workload types. And SIG Node will play a big role in this. The Sidecar KEP, which we will be talking
```

> The key challenge we will have in the next few years is how to keep innovations while maintaining
> high quality and backward compatibility of existing scenarios. SIG Node will continue to play a
> central role in Kubernetes.

Reviewer 3: ```suggestion
The key challenge we will have to face in the next few years is how to continue innovating while maintaining
high quality and backward compatibility of existing scenarios. SIG Node will continue to play a
central role in Kubernetes.
```

Reviewer: If we're quoting people, we usually keep their actual words.

Reviewer 3: Thanks, got it. Didn't know that. Even if there's a grammatical or collocation error?

Reviewer 3: Understood, thanks.

central role in Kubernetes.

> **Arpit:** And are there any ongoing research or development areas within SIG Node that excite you?

Reviewer 4: `And are there..` -- is this intentional? Or we can just begin with `Are there..`

**M/G/S:** Supporting new workload types is a fascinating area for us. Our recent exploration of
sidecar containers is a testament to this. Sidecars offer a versatile solution for enhancing
application functionality without altering the core codebase.

> **Arpit:** What are some of the challenges you've faced while maintaining SIG-Node, and how have you

Reviewer 2: ```suggestion
**Arpit:** What are some of the challenges you've faced while maintaining SIG Node, and how have you
```

overcome them?

**M/G/S:** The biggest challenge of SIG Node is its size and the many feature requests it

> receives. We are encouraging more people to join as reviewers, and are always open to improving

Reviewer 2: ```suggestion
receives. We are encouraging more people to join as reviewers and are always open to improving
```

> processes and addressing feedback. For every release, we run the feedback session at the SIG Node

Reviewer 4: `feedbacks`.

meeting and identify problematic areas and action items.

**Arpit:** Are there specific technologies or advancements that SIG Node is closely monitoring or
integrating into Kubernetes?

> **M/G/S:** Developments in components the SIG depends on, like [container

Reviewer 4: `components that the SIG depends on`

Reviewer 4: instead of `like`,  consider using`such as`

(e.g. [containerd](https://containerd.io/) and [CRI-O](https://cri-o.io/), and OS features are

> something we contribute to and monitor closely. For example, there is an upcoming `cgroup v2`

Reviewer: Not sure what was said; if the interviewees stand by “cgroup v2” (which is wrong), let's add an editor's note to explain that actually it's the v1 support that will (ie, is likely to) be deprecated.

```suggestion
something we contribute to and monitor closely. For example, there is an upcoming _cgroup v1_
```

> depreciation and removal that Kubernetes and SIG Node will need to guide Kubernetes users

Reviewer: ```suggestion
deprecation and removal that Kubernetes and SIG Node will need to guide Kubernetes users
```

through. Containerd is also releasing version `2.0`, which removes deprecated features, which will
affect Kubernetes users.

**Arpit:** Could you share a memorable experience or achievement from your time as a SIG Node
maintainer that you're particularly proud of?

> **Mathias:** I think the best moment was when my first KEP (introducing the

Reviewer 5: nit: Matthias (2xT)

> [startupProbe](https://kubebyexample.com/learning-paths/application-development-kubernetes/lesson-4-customize-deployments-application-2))

Reviewer: ```suggestion
[`startupProbe`](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#container-probes))
```

finally graduated to GA (General Availability). I also enjoy seeing my contributions being used
daily by contributors, such as the comment containing the GitHub tree hash used to retain LGTM
despite squash commits.

> ## Sidecar Containers

Reviewer: ```suggestion
## Sidecar containers
```

> ## **Sidecar Containers**

Reviewer 2: ```suggestion
## Sidecar Containers
```

**Arpit:** Can you provide more context on the concept of sidecar containers and their evolution in
the context of Kubernetes?

> **M/G/S:** The concept of [Sidecar
> Containers](https://kubernetes.io/docs/concepts/workloads/pods/sidecar-containers/) dates back to
> 2015 when Kubernetes introduced the idea of composite containers. These additional containers,

Reviewer: ```suggestion
**M/G/S:** The concept of [sidecar
containers](https://kubernetes.io/docs/concepts/workloads/pods/sidecar-containers/) dates back to
2015 when Kubernetes introduced the idea of composite containers. These additional containers,
```

  and from the primary application container.
- **Encryption and decryption:** Sidecar containers can be used to encrypt and decrypt data as it
  flows between the primary application container and external services.
- **Data synchronization:** Sidecar containers can be used to synchronize data between the primary
  application container and external databases or services.
- **Fault injection:** Sidecar containers can be used to inject faults into the primary application
  container in order to test its resilience to failures.

**Arpit:** The proposal mentions that some companies are using a fork of Kubernetes with sidecar
functionality added. Can you provide insights into the level of adoption and community interest in
this feature?

**M/G/S:** While we lack concrete metrics to measure adoption rates, the KEP has garnered
significant interest from the community, particularly among service mesh vendors like Istio, who
actively participated in its alpha testing phase. The KEP's visibility through numerous blog posts,
interviews, talks, and workshops further demonstrates its widespread appeal. The KEP addresses the

> growing demand for additional capabilities alongside main containers in Kubernetes Pods, such as

Reviewer 2: ```suggestion
growing demand for additional capabilities alongside main containers in Kubernetes pods, such as
```

network proxies, logging systems, and security measures. The community acknowledges the importance
of providing easy migration paths for existing workloads to facilitate widespread adoption of the
feature.

**Arpit:** Are there any notable examples or success stories from companies using sidecar containers
in production?

**M/G/S:** It is still too early to expect widespread adoption in production environments. The 1.29
release has only been available in Google Kubernetes Engine (GKE) since January 11, 2024, and there
still needs to be comprehensive documentation on how to enable and use them effectively via
universal injector. Istio, a popular service mesh platform, also lacks proper documentation for
enabling native sidecars, making it difficult for developers to get started with this new

> feature. However, as native sidecar support matures and documentation improves, we can expect to see

Reviewer 4: @fsmunoz - is it okay to call out Istio here? What if, by the time the blog goes live, they've actually enhanced their documentation? Instead of calling out `Istio` can we just mention that We need more precise and detailed documentation on enabling sidecars?

wider adoption of this technology in production environments.

> **Arpit:** The proposal suggests introducing a restartPolicy field for init containers to indicate

Reviewer 2: ```suggestion
**Arpit:** The proposal suggests introducing a `restartPolicy` field for init containers to indicate
```

sidecar functionality. Can you explain how this solution addresses the outlined challenges?

> **M/G/S:** The proposal to introduce a restartPolicy field for init containers addresses the

Reviewer 2: ```suggestion
**M/G/S:** The proposal to introduce a `restartPolicy` field for init containers addresses the
```

outlined challenges by utilizing existing infrastructure and simplifying sidecar management. This

> approach avoids adding new fields to the Pod specification, keeping it manageable and avoiding more

Reviewer 2: ```suggestion
approach avoids adding new fields to the pod specification, keeping it manageable and avoiding more
```

clutter. By leveraging the existing init container mechanism, sidecars can be run alongside regular
init containers during pod startup, ensuring a consistent ordering of initialization. Additionally,

> setting the restart policy of sidecar init containers to Always explicitly states that they continue

Reviewer 2: ```suggestion
setting the restart policy of sidecar init containers to `Always` explicitly states that they continue
```

running even after the main application container terminates, enabling persistent services like
logging and monitoring until the end of the workload.

**Arpit:** How will the introduction of the `restartPolicy` field for init containers affect
backward compatibility with existing Kubernetes configurations?

**M/G/S:** The introduction of the `restartPolicy` field for init containers will maintain backward
compatibility with existing Kubernetes configurations. Existing init containers will continue to
function as they have before, and the new `restartPolicy` field will only apply to init containers
explicitly marked as sidecars. This approach ensures that existing applications and deployments will
not be disrupted by the new feature, and provides a more streamlined way to define and manage
sidecars.

> # **Contributing to SIG Node**

Reviewer 2: ```suggestion
## Contributing to SIG Node
```

**Arpit:** What is the best place for the new members and especially beginners to contribute?

> New members and beginners can contribute to the Sidecar KEP by:

Reviewer 2: Was this a quote from the interviewees?

Reviewer 4: Do you think we need to expand the full form of `KEP` here? Not every newcomer would know what `KEP` stands for.
